src/utils/StkHelpers.C

```diff
@@ -0,0 +1,151 @@
+/*------------------------------------------------------------------------*/
+/*  Copyright 2019 National Renewable Energy Laboratory.                  */
+/*  This software is released under the license detailed                  */
+/*  in the file, LICENSE, which is located in the top-level Nalu          */
+/*  directory structure                                                   */
+/*------------------------------------------------------------------------*/
+
+#include "utils/StkHelpers.h"
+
+// stk_util
+#include "stk_util/parallel/ParallelReduce.hpp"
+#include "stk_util/parallel/CommSparse.hpp"
+#include "stk_util/util/SortAndUnique.hpp"
+
+namespace sierra {
+namespace nalu {
+
+void
+add_downward_relations(
+  const stk::mesh::BulkData& bulk,
+  std::vector<stk::mesh::EntityKey>& entityKeys)
+{
+  size_t numEntities = entityKeys.size();
+  for (size_t i = 0; i < numEntities; ++i) {
+    stk::mesh::Entity ent = bulk.get_entity(entityKeys[i]);
+    if (bulk.is_valid(ent)) {
+      stk::mesh::EntityRank thisRank = bulk.entity_rank(ent);
+
+      for (stk::mesh::EntityRank irank = stk::topology::NODE_RANK;
+           irank < thisRank; ++irank) {
+        unsigned num = bulk.num_connectivity(ent, irank);
+        const stk::mesh::Entity* downwardEntities = bulk.begin(ent, irank);
+
+        for (unsigned j = 0; j < num; ++j) {
+          stk::mesh::EntityKey key = bulk.entity_key(downwardEntities[j]);
+          const stk::mesh::Bucket& bkt = bulk.bucket(downwardEntities[j]);
+          if (!bkt.shared()) {
+            entityKeys.push_back(key);
+          }
+        }
+      }
+    }
+  }
+}
+
+void
+keep_elems_not_already_ghosted(
```

Comment:
@alanw0 this function seems like it doesn't do any downward relations

Reply:
Looks like you're right. The keep_only_elems function removes entities of lower rank, then the other functions figure out what needs to be added and removed, then the stk call to change_ghosting adds back in the lower-rank entities.